system/jlib/jevent.cpp

```diff
@@ -838,17 +854,297 @@ EventRecorder eventRecorder;
 
 //---------------------------------------------------------------------------------------------------------------------
 
+class CEventAttribute : public IEventAttribute
+{
+public: // IEventAttribute
+    virtual EventAttr queryId() const override
+    {
+        return id;
+    }
+
+    virtual bool isText() const override
+    {
+        switch (dataType)
+        {
+        case EATstring:
+        case EATtraceid:
+            return true;
+        }
+        return false;
+    }
+
+    virtual bool isNumeric() const override
+    {
+        switch (dataType)
+        {
+        case EATu1:
+        case EATu2:
+        case EATu4:
+        case EATu8:
+        case EATtimestamp:
+            return true;
+        }
+        return false;
+    }
+
+    virtual bool isBoolean() const override
+    {
+        if (dataType == EATbool)
+            return true;
+        return false;
+    }
+
+    virtual const char* queryTextValue() const override
+    {
+        assertex(isText());
+        return text;
+    }
+
+    virtual __uint64 queryNumericValue() const override
+    {
+        assertex(isNumeric());
+        return number;
+    }
+
+    virtual bool queryBooleanValue() const override
+    {
+        assertex(isBoolean());
+        return boolean;
+    }
+
+public:
+    void setup(EventAttr attr)
+    {
+        assertex(attr < EvAttrMax);
+        id = attr;
+        dataType = attrInformation[attr].type;
+    }
+
+    void setValue(const char* value)
+    {
+        assertex(isText());
+        text.set(value);
+    }
+
+    void setValue(__uint64 value)
+    {
+        assertex(isNumeric());
+        number = value;
+    }
+
+    void setValue(bool value)
+    {
+        assertex(isBoolean());
+        boolean = value;
+    }
+
+protected:
+    EventAttr id{EvAttrNone};
+    EventAttrType dataType{EATnone};
```

Comment:
I'm not sure this needs to be stored since it is easy to derive.  Fine to leave for the moment since it is tiny.